File: WKWebView/WKWebView/ViewController.m
//
//  ViewController.m
//  WKWebView
//
//  Created by mainone on 15/12/9.
//  Copyright © 2015年 wjn. All rights reserved.
//

#import "ViewController.h"
#import <WebKit/WebKit.h>

static void *DNWebViewContext = &DNWebViewContext;

@interface ViewController () <WKNavigationDelegate, WKUIDelegate, WKScriptMessageHandler,UIAlertViewDelegate>

@property (nonatomic, strong) WKWebViewConfiguration *configuration;//配置项
@property (nonatomic, strong) UIProgressView *progressView;//进度条
@property (nonatomic, strong) NSURL *URLToLaunchWithPermission;
@property (nonatomic, strong) UIAlertView *externalAppPermissionAlertView;
@property (nonatomic, strong) WKBackForwardList *BFList;//网页加载记录

@end

@implementation ViewController

- (WKWebView *)webView {
    if (!_webView) {
        _webView = [[WKWebView alloc]initWithFrame:self.view.bounds configuration:self.configuration];
        _webView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [_webView addObserver:self forKeyPath:NSStringFromSelector(@selector(estimatedProgress)) options:0 context:NULL];//监听进度条变化
        [_webView addObserver:self forKeyPath:@"title" options:NSKeyValueObservingOptionNew context:nil];
        _webView.navigationDelegate = self;
        _webView.UIDelegate = self;
        _webView.multipleTouchEnabled = YES;
        _webView.autoresizesSubviews = YES;
        _webView.scrollView.alwaysBounceVertical = YES;
        [self.view addSubview:_webView];
    }
    return _webView;
}

//偏好设置
- (WKWebViewConfiguration *)configuration {
    if (!_configuration) {
        _configuration = [[WKWebViewConfiguration alloc]init];
    }
    return _configuration;
}

- (UIProgressView *)progressView {
    if (!_progressView) {
        _progressView = [[UIProgressView alloc]initWithProgressViewStyle:UIProgressViewStyleDefault];
        _progressView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleTopMargin;
        _progressView.trackTintColor = [UIColor whiteColor];
        _progressView.progressTintColor = [UIColor orangeColor];
        _progressView.frame = CGRectMake(0, self.navigationController.navigationBar.frame.size.height-self.progressView.frame.size.height, self.view.frame.size.width, _progressView.frame.size.height);
    }
    return _progressView;
}

- (UIAlertView *)externalAppPermissionAlertView {
    if (!_externalAppPermissionAlertView) {
        _externalAppPermissionAlertView = [[UIAlertView alloc] initWithTitle:@"提示" message:@"这网页试图打开外部应用。你确定你想要打开它吗?" delegate:self cancelButtonTitle:@"取消" otherButtonTitles:@"前往", nil];
    }
    return _externalAppPermissionAlertView;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    
    if ([keyPath isEqualToString:NSStringFromSelector(@selector(estimatedProgress))] && object == self.webView) {
        [self.progressView setAlpha:1.0f];
        BOOL animated = self.webView.estimatedProgress > self.progressView.progress;
        [self.progressView setProgress:self.webView.estimatedProgress animated:animated];
        
        if(self.webView.estimatedProgress >= 1.0f) {
            [UIView animateWithDuration:0.3f delay:0.3f options:UIViewAnimationOptionCurveEaseOut animations:^{
                [self.progressView setAlpha:0.0f];
            } completion:^(BOOL finished) {
                [self.progressView setProgress:0.0f animated:NO];
            }];
        }
    }else if ([keyPath isEqualToString:@"title"]) {
        self.title = self.webView.title;
    }else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
    
}

- (void)dealloc {
    [self.webView setNavigationDelegate:nil];
    [self.webView setUIDelegate:nil];
    if ([self isViewLoaded]) {
        [self.webView removeObserver:self forKeyPath:NSStringFromSelector(@selector(estimatedProgress))];
        [self.webView removeObserver:self forKeyPath:@"title"];
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    NSURL *url = [NSURL URLWithString:@"http://www.baidu.com"];
    [self.webView loadRequest:[NSURLRequest requestWithURL:url]];
    
    UIBarButtonItem *closeBtn = [[UIBarButtonItem alloc] initWithTitle:@"后退" style:UIBarButtonItemStylePlain target:self action:@selector(closeAction)];
    self.navigationItem.leftBarButtonItem = closeBtn;
}

//可以前进或后退任意一级
- (void)closeAction {
    NSLog(@"我其实是返回");
    WKBackForwardList *backForwardList = self.webView.backForwardList;
    if (backForwardList.backList.count > 0) {
        WKBackForwardListItem *item = [backForwardList itemAtIndex:-1];
        NSLog(@"%@-----%@",item.title,item.URL);
        [self.webView goToBackForwardListItem:item];
    }else {
        [self.navigationController popViewControllerAnimated:YES];
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:YES];
    [self.navigationController.navigationBar addSubview:self.progressView];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:YES];
    [self.progressView removeFromSuperview];
}

#pragma mark - 外部链接
//外部应用程序需要打开URL
- (BOOL)externalAppRequiredToOpenURL:(NSURL *)URL {
    NSSet *validSchemes = [NSSet setWithArray:@[@"http", @"https"]];
    return ![validSchemes containsObject:URL.scheme];
}

- (void)launchExternalAppWithURL:(NSURL *)URL {
    self.URLToLaunchWithPermission = URL;
    if (![self.externalAppPermissionAlertView isVisible]) {
        [self.externalAppPermissionAlertView show];
    }
}

#pragma mark - UIAlertViewDelegate
- (void)alertView:(UIAlertView *)alertView didDismissWithButtonIndex:(NSInteger)buttonIndex {
    if(alertView == self.externalAppPermissionAlertView) {
        if(buttonIndex != alertView.cancelButtonIndex) {
            [[UIApplication sharedApplication] openURL:self.URLToLaunchWithPermission];
        }
        self.URLToLaunchWithPermission = nil;
    }
}

- (void)loadURL:(NSURL *)URL {
    [self.webView loadRequest:[NSURLRequest requestWithURL:URL]];
}

#pragma mark - WKNavigationDelegate
//发送请求之前，决定是否跳转
- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler {
    NSLog(@"发送请求之前，决定是否跳转");
    NSURL *URL = navigationAction.request.URL;
    if(![self externalAppRequiredToOpenURL:URL]) {
        if(!navigationAction.targetFrame) {
            [self loadURL:URL];
            decisionHandler(WKNavigationActionPolicyCancel);
            return;
        }
    }else if([[UIApplication sharedApplication] canOpenURL:URL]) {
        [self launchExternalAppWithURL:URL];
        decisionHandler(WKNavigationActionPolicyCancel);
        return;
    }

    decisionHandler(WKNavigationActionPolicyAllow);
}
//收到响应后，决定是否跳转
- (void)webView:(WKWebView *)webView decidePolicyForNavigationResponse:(WKNavigationResponse *)navigationResponse decisionHandler:(void (^)(WKNavigationResponsePolicy))decisionHandler {
    NSLog(@"收到响应后，决定是否跳转");
    decisionHandler(WKNavigationResponsePolicyAllow);
}
//接收到服务器跳转请求之后调用
- (void)webView:(WKWebView *)webView didReceiveServerRedirectForProvisionalNavigation:(null_unspecified WKNavigation *)navigation {
    NSLog(@"接收到服务器跳转请求之后调用");
}
//开始加载
- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(null_unspecified WKNavigation *)navigation {
    NSLog(@"开始加载");
}
//加载失败
- (void)webView:(WKWebView *)webView didFailProvisionalNavigation:(null_unspecified WKNavigation *)navigation withError:(NSError *)error {
    NSLog(@"加载失败");
}
//内容开始返回
- (void)webView:(WKWebView *)webView didCommitNavigation:(null_unspecified WKNavigation *)navigation {
    NSLog(@"内容开始返回");
}
//加载完成
- (void)webView:(WKWebView *)webView didFinishNavigation:(null_unspecified WKNavigation *)navigation {
    NSLog(@"加载完成");
}
//接收内容时发生错误
- (void)webView:(WKWebView *)webView didFailNavigation:(null_unspecified WKNavigation *)navigation withError:(NSError *)error {
    NSLog(@"接收内容时发生错误");
}
////身份验证
//- (void)webView:(WKWebView *)webView didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition disposition, NSURLCredential *__nullable credential))completionHandler {
//}
//终止页面加载
- (void)webViewWebContentProcessDidTerminate:(WKWebView *)webView NS_AVAILABLE(10_11, 9_0) {
    NSLog(@"终止页面加载");
}

#pragma mark - WKUIDelegate
//创建一个新的WebView
- (nullable WKWebView *)webView:(WKWebView *)webView createWebViewWithConfiguration:(WKWebViewConfiguration *)configuration forNavigationAction:(WKNavigationAction *)navigationAction windowFeatures:(WKWindowFeatures *)windowFeatures {
    if (!navigationAction.targetFrame.isMainFrame) {
        [webView loadRequest:navigationAction.request];
    }
    return nil;
}
//通知应用程序正常关闭
- (void)webViewDidClose:(WKWebView *)webView NS_AVAILABLE(10_11, 9_0) {
    
}
//弹出警告框时
- (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSString *)message initiatedByFrame:(WKFrameInfo *)frame completionHandler:(void (^)(void))completionHandler {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"警告框" message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        completionHandler();
    }];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
}
//弹出确认框
- (void)webView:(WKWebView *)webView runJavaScriptConfirmPanelWithMessage:(NSString *)message initiatedByFrame:(WKFrameInfo *)frame completionHandler:(void (^)(BOOL result))completionHandler {
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"确认框" message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *sureAction = [UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        completionHandler(YES);
    }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
        completionHandler(NO);
    }];
    [alert addAction:sureAction];
    [alert addAction:cancelAction];
    
    [self presentViewController:alert animated:YES completion:nil];
}
//弹出输入框
- (void)webView:(WKWebView *)webView runJavaScriptTextInputPanelWithPrompt:(NSString *)prompt defaultText:(nullable NSString *)defaultText initiatedByFrame:(WKFrameInfo *)frame completionHandler:(void (^)(NSString * __nullable result))completionHandler {
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:prompt message:defaultText preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.textColor = [UIColor blackColor];
    }];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"完成" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        completionHandler(alert.textFields[0].text);
    }];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
}

#pragma mark - WKScriptMessageHandler
//从web界面中接收到一个脚本时调用
- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message {
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

@end

File: WKWebView_Swift/WKWebView_Swift/ViewController.swift
//
//  ViewController.swift
//  WKWebView_Swift
//
//  Created by mainone on 16/4/25.
//  Copyright © 2016年 wjn. All rights reserved.
//

import UIKit
import WebKit


class ViewController: UIViewController, WKScriptMessageHandler, WKUIDelegate, WKNavigationDelegate, UIAlertViewDelegate {
    
    var URLToLaunchWithPermission: NSURL!
    
    
    deinit {
        self.webview.removeObserver(self, forKeyPath: "loading")
        self.webview.removeObserver(self, forKeyPath: "title")
        self.webview.removeObserver(self, forKeyPath: "estimatedProgress")
        
        self.webview.removeFromSuperview()
        self.progressView.removeFromSuperview()
        self.webview.UIDelegate = nil
        self.webview.navigationDelegate = nil
        self.alertView.delegate = nil
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        //加载网页
        self.webview.loadRequest(NSURLRequest(URL:NSURL(string: "http://m.baidu.com")!))
        
        
        //监听属性
        self.webview.addObserver(self, forKeyPath: "loading", options: .New, context: nil)
        self.webview.addObserver(self, forKeyPath: "title", options: .New, context: nil)
        self.webview.addObserver(self, forKeyPath: "estimatedProgress", options: .New, context: nil)
        
        //设置前进后退
        let leftItem = UIBarButtonItem(title: "后退", style: .Done, target: self, action: #selector(ViewController.previousPage))
        
        let closeItem = UIBarButtonItem(title: "关闭", style: .Done, target: self, action: #selector(ViewController.closePage))
        self.navigationItem.leftBarButtonItems = Array(arrayLiteral: leftItem, closeItem)
        
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "前进", style: .Done, target: self, action: #selector(ViewController.nextPage))
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    //MARK: - KVO
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if keyPath == "estimatedProgress" {
            self.progressView.alpha = 1.0
            let animated: Bool = Float(self.webview.estimatedProgress) > self.progressView.progress
            self.progressView.setProgress(Float(self.webview.estimatedProgress), animated: animated);
            if self.webview.estimatedProgress >= 1.0 {
                UIView.animateWithDuration(0.3, delay: 0.3, options: .CurveEaseInOut, animations: { 
                    self.progressView.alpha = 0.0
                    }, completion: { (finished) in
                        self.progressView.setProgress(0.0, animated: false)
                })
            }
        } else if keyPath == "title" {
            self.title = self.webview.title;
        } else if keyPath == "loading" {
            print("loading");
        }
    }
    
    
    //MARK: - WKScriptMessageHandler
    
    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
       
    }
    
    //MARK: - WKNavigationDelegate
    
    //决定导航的动作，通常用于处理跨域的链接是否导航，webkit对跨域进行安全检查，不允许出现跨域
    func webView(webView: WKWebView, decidePolicyForNavigationAction navigationAction: WKNavigationAction, decisionHandler: (WKNavigationActionPolicy) -> Void) {
        
        let URL: NSURL = navigationAction.request.URL!
        if !self.externalAppRequiredToOpenURL(URL) {
            if (navigationAction.targetFrame == nil) {
                self.webview.loadRequest(NSURLRequest(URL: URL))
                decisionHandler(.Cancel)
                return
            }
        } else if UIApplication.sharedApplication().canOpenURL(URL) {
            self.launchExternalAppWithURL(URL)
            decisionHandler(.Cancel)
            return
        }
        
         decisionHandler(.Allow);
    }
    
    //决定是否允许导航相应，如果不允许就不会跳转到该链接的页面
    func webView(webView: WKWebView, decidePolicyForNavigationResponse navigationResponse: WKNavigationResponse, decisionHandler: (WKNavigationResponsePolicy) -> Void) {
        decisionHandler(.Allow)
    }
    
    //页面内容开始加载
    func webView(webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        
    }
    
    //接收到服务器跳转请求之后调用
    func webView(webView: WKWebView, didReceiveServerRedirectForProvisionalNavigation navigation: WKNavigation!) {
       
    }
    
    //网页内容加载失败
    func webView(webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: NSError) {
        
    }
    
    //页面内容到达mainFrame的时候调用（若想在mainFrame中注入js代码也可在此方法中添加）
    func webView(webView: WKWebView, didCommitNavigation navigation: WKNavigation!) {
        
    }
    
    //页面内容完成加载
    func webView(webView: WKWebView, didFinishNavigation navigation: WKNavigation!) {
       
    }
    
    //接收内容发生错误
    func webView(webView: WKWebView, didFailNavigation navigation: WKNavigation!, withError error: NSError) {
       
    }
    
    //若我们的请求需要授权、证书等，需处理此方法
    func webView(webView: WKWebView, didReceiveAuthenticationChallenge challenge: NSURLAuthenticationChallenge, completionHandler: (NSURLSessionAuthChallengeDisposition, NSURLCredential?) -> Void) {
        
         completionHandler(.PerformDefaultHandling, nil)
    }
    
//    //终止页面加载 (iOS 9.0 )
//    func webViewWebContentProcessDidTerminate(webView: WKWebView) {
//        print(#function)
//    }
    
    //MARK: - WKUIDelegate
    
    //创建一个新的WebVIew
    func webView(webView: WKWebView, createWebViewWithConfiguration configuration: WKWebViewConfiguration, forNavigationAction navigationAction: WKNavigationAction, windowFeatures: WKWindowFeatures) -> WKWebView? {
        return nil
    }
    
    //通知应用程序正常关闭（iOS 9.0）
    func webViewDidClose(webView: WKWebView) {
        
    }
    
    //弹出警告框
    func webView(webView: WKWebView, runJavaScriptAlertPanelWithMessage message: String, initiatedByFrame frame: WKFrameInfo, completionHandler: () -> Void) {
        
    }
    
    //弹出确认框
    func webView(webView: WKWebView, runJavaScriptConfirmPanelWithMessage message: String, initiatedByFrame frame: WKFrameInfo, completionHandler: (Bool) -> Void) {
        
    }
    
    //弹出输入框
    func webView(webView: WKWebView, runJavaScriptTextInputPanelWithPrompt prompt: String, defaultText: String?, initiatedByFrame frame: WKFrameInfo, completionHandler: (String?) -> Void) {
        
    }
    
    //MARK: - UIAlertViewDelegate
    
    func alertView(alertView: UIAlertView, didDismissWithButtonIndex buttonIndex: Int) {
        if alertView == self.alertView {
            if buttonIndex != alertView.cancelButtonIndex {
                UIApplication.sharedApplication().openURL(self.URLToLaunchWithPermission)
            }
            self.URLToLaunchWithPermission = nil
        }
    }
    
    //MARK: - 方法
    
    func previousPage() {
        if self.webview.canGoBack {
            self.webview.goBack()
        }
    }
    
    func nextPage() {
        if self.webview.canGoForward {
            self.webview.goForward()
        }
    }
    
    func closePage() {
        let backForwardList = self.webview.backForwardList
        let index = backForwardList.backList.count
        if index > 0 {
            let item = backForwardList.itemAtIndex(-index)
            self.webview.goToBackForwardListItem(item!)
        }else {
            
        }
    }
    
    //外部应用程序打开URL
    func externalAppRequiredToOpenURL(URL: NSURL) -> Bool {
        let validSchemes = NSSet(arrayLiteral: "http", "https")
        return !validSchemes.containsObject(URL.scheme)
    }
    
    func launchExternalAppWithURL(URL: NSURL) {
        self.URLToLaunchWithPermission = URL
        self.alertView.show()
        
    }
    
    //MARK: - 初始化
    
    lazy var webview: WKWebView = {
        let webV = WKWebView(frame: self.view.bounds, configuration: self.configuretion)
        webV.autoresizingMask = [.FlexibleWidth, .FlexibleHeight]
        self.view.addSubview(webV)
        webV.navigationDelegate = self
        webV.UIDelegate = self
        print("我创建了你")
        return webV
    }()
    
    lazy var configuretion: WKWebViewConfiguration = {
        let configuretion = WKWebViewConfiguration()
        
        //webview的偏好设置
        configuretion.preferences = WKPreferences()
        configuretion.preferences.minimumFontSize = 10
        configuretion.preferences.javaScriptEnabled = true
        
        //默认不通过js自动打开窗口的，必须用户交互才能打开
        configuretion.preferences.javaScriptCanOpenWindowsAutomatically = false
        
        //通过js与webview内容交互配置
        configuretion.userContentController = WKUserContentController()
        return configuretion
    }()
    
    lazy var progressView: UIProgressView = {
        let progressV = UIProgressView(progressViewStyle: .Default)
        progressV.autoresizingMask = [.FlexibleWidth, .FlexibleHeight, .FlexibleTopMargin]
        progressV.frame = CGRectMake(0, self.navigationController!.navigationBar.frame.size.height - progressV.frame.size.height, self.view.frame.size.width, progressV.frame.size.height)
        progressV.trackTintColor = UIColor.whiteColor()
        progressV.progressTintColor = UIColor.orangeColor()
        self.navigationController!.navigationBar.addSubview(progressV)
        return progressV;
    }()

    lazy var alertView: UIAlertView = {
        let alertV = UIAlertView(title: "提示", message: "这网页试图打开外部应用。你确定你想要打开它吗?", delegate: self, cancelButtonTitle: "取消", otherButtonTitles: "前往")
        alertV.delegate = self
        return alertV
    }()
    
}

